Migrate `block-lab` namespace in `post_content`

kienstra opened the pull request:
<!--- Please summarize this PR in the title above -->

#### Changes
* Adds PHP functions to migrate the entire site's `post_content` from the `block-lab/` block namespace to `genesis-custom-blocks/`
* There are more considerations after this PR, like a REST API or AJAX endpoint that the UI can use

#### Testing instructions
1. `wp db export # if you'd like, so you can restore to this backup of your DB `
2. `wp eval-file tests/bin/migrate-post-content.php`

<img width="546" alt="wp-cli-file" src="https://user-images.githubusercontent.com/4063887/84827506-2ff4c580-afea-11ea-9081-f282c205f8cf.png">

3. Expected: The WP-CLI command results in 'Success', like above
4. Go to the 'Revision' links shown above
5. Expected: The revision diff only involves `block-lab`. Nothing else should have changed:
<img width="1020" alt="exampled-diff-here" src="https://user-images.githubusercontent.com/4063887/84827765-8feb6c00-afea-11ea-893c-c7671c91d9f5.png">

6. Expected: If there's no `block-lab/` namespace in the content, the post should not be shown in the WP-CLI command output above, and there should of course be no revision.
7. It'd be great if you could go to 40-50 of those revision URLs to sanity-check them, assuming there are that many. It's tedious, but they're all just one revision ID higher than the next. 

8. To reverse this migration:
```sh
wp shell
( new Block_Lab\Blocks\Migration\Post_Content( 'genesis-custom-blocks', 'block-lab' ) )->migrate_all();
```

kienstra committed:
Begin migration of post_content with Block Lab blocks

This is only the start,
there are more considerations like a REST API
or ajax endpoint that the UI can use.

kienstra committed:
Correct a param tag in the DocBlock of migrate_single()

kienstra committed:
Add more test cases for migrating post content

Including non-Block Lab blocks
and multiple Block Lab blocks.

kienstra committed:
Migrate the post_content to the new namespace

Query for all of the blocks that probably have
a Block Lab block, by looking for wp:block-lab/.
Then, pass those blocks to a function calls preg_replace()
to change the namespace to genesis-custom-blocks.

kienstra committed:
Add a test for an invalid post ID passed to migrate_single()

If there's no post for this,
it should return a WP_Error.

kienstra committed:
Move ->assertEmpty() out of a foreach loop, use array_diff()

kienstra committed:
If no post_content was replaced in preg_replace, don't update the post

Instead, return a WP_Error,
as there's no need to update it.

kienstra committed:
Remove needless import of WP_Post

kienstra committed:
Clarify DocBlock of query_for_posts()

Remove an extra 'The',
and delete a lint that was confusing.

kienstra committed:
Change the domain of the testing site

kienstra committed:
Fix the name of some functions in covers tags

There is no reset_block_row(),
only reset_block_rows().

kienstra committed:
Add more covers tags, for private helper functions

query_for_posts() is only called internally,
so add covers tags for them.

kienstra committed:
Minor space edits, like adding an empty line

kienstra committed:
Prevent overwriting revisions

This will keep the revision history of the post.
The migration will just be the latest revision.
This will allow debugging.
If there was a problem, they can look at the revision.
And we can test this by looking at them

kienstra committed:
Add a test-only WP-CLI script to test the migration

Prints a table of the posts that were migrated,
and the errors.

kienstra commented on tests/php/integration/test-repeater-template-output.php:
```diff
@@ -182,9 +182,9 @@ public function get_block_config() {
-	 * @covers \block_row_value()
+	 * @covers \reset_block_rows()
+	 * @covers \block_sub_field()
+	 * @covers \block_sub_value()
```
I referenced functions that didn't exist 😄

kienstra commented on tests/bin/migrate-post-content.php:
```diff
@@ -0,0 +1,116 @@
+/**
+ * Migrates the entire site's post content to the new namespace.
+ *
+ * For testing, do not use in production.
```
This WP-CLI script is mainly for our own testing. I don't think we should publicize this. 

It also checks that you really want to migrate:

<img width="544" alt="sure-migr" src="https://user-images.githubusercontent.com/4063887/84828910-70554300-afec-11ea-873e-ba8321d9d27a.png">

kienstra commented on php/blocks/migration/class-post-content.php:
```diff
@@ -0,0 +1,131 @@
+		$query_limit = 10;
+		return $wpdb->get_results(
+			$wpdb->prepare(
+				"SELECT * FROM {$wpdb->posts} WHERE post_type != %s AND post_content LIKE %s LIMIT %d",
```
Like in #588, this shouldn't get posts that were already migrated. 

Their `post_content` should have changed to have the new namespace, so `post_content LIKE %s` shouldn't be satisfied.

kienstra replied:
This queries for posts that have `wp:block-lab/` in the content. I don't think that's possible with `WP_Query`.

kienstra commented on php/blocks/migration/class-post-content.php:
```diff
@@ -0,0 +1,131 @@
+		return $wpdb->get_results(
+			$wpdb->prepare(
+				"SELECT * FROM {$wpdb->posts} WHERE post_type != %s AND post_content LIKE %s LIMIT %d",
+				'revision',
```
This doesn't migrate `revision` posts. I thought that'd be like overwriting their 'GIT history' of posts.

This migration creates a new revision for a migrated post:

<img width="825" alt="rev-here" src="https://user-images.githubusercontent.com/4063887/84831024-325a1e00-aff0-11ea-9aae-0e719dcfd03b.png">

This should make it easier for us to debug if something went wrong.

lukecarbis replied:
Great thinking.

kienstra commented:
Hi @lukecarbis and @RobStino,
Hope your day's off to a great start.

Could you please review this when you have a chance? The [top comment](https://github.com/getblocklab/block-lab/pull/606#issue-432064433) is up to date.

Thanks!

kienstra commented:
For reviewers in an environment without any Block Lab block:

1. Publish the 'Example Block':

<img width="842" alt="publish-example-block" src="https://user-images.githubusercontent.com/4063887/85158584-2190ee80-b222-11ea-8e0a-9582314af50d.png">

2. Create a few posts, and add that block to it. 
3. If you're adventurous, create a few more Block Lab blocks like the 'Example Block', and add them to posts.

kienstra committed:
Merge branch 'add/full-migration' into add/migrate-post-content

lukecarbis commented on php/blocks/migration/class-post-content.php:
```diff
@@ -0,0 +1,131 @@
+	public function migrate_single( $post_id ) {
+		$post = get_post( $post_id );
+		if ( ! isset( $post->ID ) ) {
+			return new WP_Error( 'Invalid post ID' );
```
`WP_Error` has an error "code" as the first parameter of its constructor. The message comes second. [Here's an example](https://github.com/WordPress/WordPress/blob/a576a13246f85dfa76183911041d666a5cd42818/wp-includes/rest-api/endpoints/class-wp-rest-users-controller.php#L175).

Also, should we translate error messages?

My suggestion is something like:
```PHP
return new WP_Error(
	'invalid_post_id',
	__( 'Invalid post ID', 'genesis_custom_blocks' )
);
```

kienstra replied:
Good suggestion, the message should be second.

kienstra replied:
Thanks for suggesting this, https://github.com/getblocklab/block-lab/pull/606/commits/d708186bac9981ec3fc67f0f842ee4a6c0866d31 uses your suggestion.

lukecarbis commented on php/blocks/migration/class-post-content.php:
```diff
@@ -0,0 +1,131 @@
+		);
+
+		if ( 0 === $replacement_count ) {
+			return new WP_Error( 'Post content did not have blocks with the namespace' );
```
See `WP_Error` comment above.

kienstra replied:
Good idea

kienstra replied:
Thanks, fixed in https://github.com/getblocklab/block-lab/pull/606/commits/d708186bac9981ec3fc67f0f842ee4a6c0866d31

lukecarbis commented on php/blocks/migration/class-post-content.php:
```diff
@@ -0,0 +1,131 @@
+			$posts = $this->query_for_posts();
+		}
+
+		return $results;
```
I think I found a small problem with this looping `while` approach.

Suppose 1 post doesn't migrate. It returns a `WP_Error` due to an invalid ID, for example, or the `wp_update_post()` function breaks.

Next loop through, that same post that caused an error in the previous loop will be reattempted, since it will be the first result in `query_for_posts`. That's no big problem. But what if it breaks every time? What if the second and third and fourth time we try to update it, it always returns the same error?

In that case, it's still no big deal, it's just slightly less efficient, as we'll only be migrating 9 posts with every loop through instead of 10.

BUT… what if it's not 1 post, but all 10 posts! Or 5 posts in the first loop through and 5 posts in the second (causing a total of 10 errors after the second loop). Since our `$query_limit` is set to 10, we'll end up with an infinite loop.

Having said that, I think the fix for that problem potentially introduces a ton of complexity. Happy to move forward with this PR knowing that this super unlikely edge case exists, but not fixing it.

kienstra replied:
That's a good catch, I'll think about that. It could definitely create an infinite loop if there are 10 failures.

kienstra replied:
Really good point, I think that is a case we should handle.

What do you think about https://github.com/getblocklab/block-lab/pull/606/commits/f6aada67ce83125c46285ed04cce24375ed3389c?

It returns from `migrate_all()` if there are more than 20 `WP_Error`s in migration.

It'll prevent an infinite loop. 

If there was a `WP_Error`, something might have gone very wrong. Like `wp_insert_post()` [tried to insert an empty post](https://github.com/WordPress/wordpress-develop/blob/5e2ef3d93739610e97dd20f5f254a355a4eb5bd3/src/wp-includes/post.php#L3697). If there are over 20 of those, it's probably better to exit. 

It might be possible to keep track of failed post IDs and exclude them from the SQL query, maybe with `NOT IN`. But that might not be very efficient.

lukecarbis replied:
Yeah, this is all pretty edge case. I'm happy with the 20 errors limit. 🚢

kienstra replied:
Nice, thanks!

lukecarbis commented on tests/php/unit/blocks/migration/test-class-post-content.php:
```diff
@@ -0,0 +1,280 @@
+	 *
+	 * @covers \Block_Lab\Blocks\Migration\Post_Content::migrate_single()
+	 */
+	public function test_migrate_single_invalid_post_id() {
```
Great idea.

kienstra replied:
Thanks!

lukecarbis commented:
Looks like e2e tests are failing?

lukecarbis reviewed (approved):
Nice work on this Ryan.

Migration script worked beautifully.

kienstra committed:
Proposal: exit after 20 errors in migrating posts

If there were somehow 20 errors,
something could have gone very wrong.
For example, the post is somehow empty.
This can at least prevent an infinite loop.

kienstra commented:
>Looks like e2e tests are failing?

Yeah, that's a flaky test case that I introduced. I'll have to fix it some time, whether it passes now or not.

kienstra commented on php/blocks/migration/class-post-content.php:
```diff
@@ -0,0 +1,137 @@
+			foreach ( $posts as $post ) {
+				if ( isset( $post->ID ) ) {
+					$migrated_post = $this->migrate_single( $post->ID );
+					$results[]     = $migrated_post;
```
Before releasing this, we might remove `$results`, and only return `WP_Error` messages if there were any. If there are hundreds of posts migrated, this could be too big.

But for now, it lets us test this and enables the WP-CLI output of exactly what posts were migrated.

kienstra committed:
Use correct arguments for WP_Error

Use a simply key for the first arguments,
and a translated string for the second

kienstra commented:
This PR can still be tested with the same [testing instructions](https://github.com/getblocklab/block-lab/pull/606#issue-432064433), just checkout the [add/full-migration](https://github.com/getblocklab/block-lab/tree/add/full-migration) branch.

Here are some [instructions](https://github.com/getblocklab/block-lab/pull/606#issuecomment-646743052) to populate Block Lab blocks if needed.

It'd be interesting to know roughly how long `wp eval-file tests/bin/migrate-post-content.php` takes in your local. 2-3 seconds, or 20 seconds?
